### ISSUE TITLE
[scan] D9 delegates categorical design factors to BayBE and the handback breaks at the cost head — the conformal path takes an augmented design matrix, `design_context` unpacks exactly five columns, so a carbon-source campaign gets a calibrated interval and no $/kg

### ISSUE THREAD

**Issue by dcbaugher** [enginbio/engin-suite#396]
**Lens:** Breadth (rotation idx 1) · **Scanned:** 2026-09-11

## What

Engin's answer to a qualitative design factor — which carbon source, which nitrogen
source, which media base, which inducer — is `D9`: don't build mixed-variable BO,
compose with something that has it. `AGENTS.md:71-74` states the mechanism outright:

> `engin_core.fit_gp` is for low-dimensional continuous design spaces. … Bring a
> different estimator and feed its `(mean, sd)` to the shared conformal and
> acquisition primitives, which are estimator-agnostic.

`docs/ecosystem.md` routes accordingly: BayBE for "mixed discrete/continuous/categorical
parameters" (`:172-174`), BoFire for "constrained mixed continuous/discrete/categorical
spaces" (`:229-233`).

**That delegation works for calibration and fails at cost.** The conformal and
acquisition primitives really are estimator-agnostic —
`split_conformal_multiplier(y_cal, mean_cal, sd_cal)` (`gp.py:399`),
`expected_improvement(mean, sd, best)` (`recommend.py:21`) and
`prob_at_least(mean, sd, target)` (`gp.py:529`) are pure arrays and do not care how
many columns the design matrix had. An augmented `(n, 5 + k)` matrix with a one-hot or
descriptor-encoded carbon source fits, calibrates and recommends. I ran it.

The cost half does not. `CostModel` is declared as the interchangeable seam —
`cost_per_kg(self, titer_g_L, U)` (`tea.py:425`) — but the default implementation
binds `U` to the five simulator knobs by position:

```python
# tea.py:168, inside design_context()
feed_rate, feed_start, Sf, _induction, S0 = (phys[:, i] for i in range(5))
```

`design_context` calls `unit_to_physical` (`simulator.py:172-180`), which does
`lo + U * (hi - lo)` against length-5 bound arrays. Anything but five columns raises a
broadcast error rather than a message. So `ParametricCostModel.cost_per_kg`
(`tea.py:446-449`) — and therefore `cost_samples`, `cost_summary`,
`expected_cost_reduction` and `recommend_batch_by_cost` — is unreachable for the design
matrix the composed search produced.

`recommend_batch_by_cost` has a second, quieter version of the same problem:
`candidates = rng.random((pool, len(gp.ell)))` (`tea.py:824`). It samples the
candidate pool as a *continuous* unit cube of the GP's ARD dimension, so an augmented
one-hot block is drawn as continuous noise — fractional category memberships that
correspond to no carbon source.

`BioSteamCostModel.cost_per_kg` (`tea.py:529`) is shape-agnostic, because it hands `U`
to a caller-supplied `build_model`. So the escape hatch is the one that needs the
`[tea]` extra and a flowsheet.

The delegation has also never been exercised on the thing it delegates:
`benchmarks/baybe_baseline.py:67,77` imports exactly one parameter type,
`NumericalContinuousParameter`, and builds `SearchSpace.from_product` over the same
five continuous knobs. The BayBE comparison is BayBE doing the job `fit_gp` already
does.

## Why it matters

`D13` says the objective is net $/kg, not titer, and the suite's stated contribution is
a *calibrated posterior driving a cost decision* — `tea.py`'s own docstring calls that
seam "what is genuinely underexplored". For a campaign with a qualitative factor, a
user who follows `D9` and `AGENTS.md` to the letter gets the calibrated interval and
loses the objective. They land on titer, which is the metric `D13` exists to argue
against.

The decision is not exotic. The classical fermentation-optimization literature screens
qualitative factors first and screens them widest, then freezes them by OFAT before any
model-based design runs: carbon source 8 levels / nitrogen 9 / inorganic salt 7
(*AMB Express* 15:145, 2025); 7/7/7 (*Front. Microbiol.* 16:1552645, 2025); 5/4-5/5
(*Microb. Cell Fact.* 24:173, 2025). The largest-effect decisions in those papers are
made by the least rigorous method available and never enter the design space. That is
the unserved decision, and it is unserved in the field, not only here.

The one 2025 wet-lab campaign that kept the categorical factor *inside* the loop is
Narayanan et al., and it found the choice is not separable from the rest: the optimal
co-feed identity differed per target protein, and "no two molecules converged to the
same composition". Running one campaign per level — the workaround — is exactly the
thing that paper's result says loses information.

Magnitude, stated honestly: this blocks the cost head for a class of campaign Engin has
never run and has no user for. The consequence is about what the composition story
promises, not about a number anyone has published.

## Evidence

- `packages/engin-core/src/engin_core/tea.py:168` — `design_context` unpacks exactly
  five columns. `tea.py:425` — the `CostModel` protocol that is supposed to make the
  cost backend interchangeable. `tea.py:446-449` — `ParametricCostModel.cost_per_kg`
  calls `design_context`. `tea.py:824` — the cost recommender's candidate pool is
  `rng.random((pool, len(gp.ell)))`.
- `packages/engin-core/src/engin_core/simulator.py:172-180` — `unit_to_physical` maps
  the unit cube against length-5 bound arrays. `simulator.py:153-155` —
  `ReactorConfig` raises if `knob_bounds` does not cover exactly `KNOB_NAMES`. (That
  one is correct and not the defect: the knob set is fixed by the ODEs.)
- `packages/engin-core/src/engin_core/gp.py:399,529` and `recommend.py:21` — the
  primitives `AGENTS.md` calls estimator-agnostic, and they are. Measured: a
  `(n, 5+k)` augmented matrix fits, yields a conformal multiplier, and scores EI.
- `packages/engin-core/benchmarks/baybe_baseline.py:67,77` — the composed baseline
  uses `NumericalContinuousParameter` only.
- `packages/engin-core/src/engin_core/loaders.py:584-596` — `load_timeseries` drops
  non-numeric columns, at `warning` level with code `non-numeric-columns-dropped`.
  **Not a barrier**, and I originally had it as one: `load_endpoints`
  (`loaders.py:522-546`) carries unmapped columns through untouched, and an endpoint
  DoE table is the right shape for this data anyway.
- Narayanan, Hinckley, Barry, Dang, Wolffe, Atari, Tseng & Love, *Accelerating cell
  culture media development using Bayesian optimization-based iterative experimental
  design*, **Nat Commun 16:6055 (2025)**,
  [doi:10.1038/s41467-025-61113-5](https://doi.org/10.1038/s41467-025-61113-5).
  Verbatim: "the type of carbon co-feed supplement in production (categorical
  factor)"; "To account for the categorical variables in our surrogate GP model, we
  adopted a modified kernel design instead of using OHE"; "the average errors made by
  GP with the designed kernel were 33−50% smaller compared to OHE"; "no two molecules
  converged to the same composition". Already raised in a different frame on
  [#357](https://github.com/enginbio/engin-suite/issues/357), which notes
  `Carbon_Type` is categorical with ~19 levels and that `fit_gp` cannot consume it
  without an encoding decision nobody has made. This issue is the *cost-head* half
  that observation does not reach.
- OFAT-then-freeze pattern, three 2025 papers, each verified by fetching the article:
  [doi:10.1186/s13568-025-01912-7](https://doi.org/10.1186/s13568-025-01912-7) (8/9/7
  levels), [doi:10.3389/fmicb.2025.1552645](https://doi.org/10.3389/fmicb.2025.1552645)
  (7/7/7), [doi:10.1186/s12934-025-02798-7](https://doi.org/10.1186/s12934-025-02798-7)
  (5/4-5/5).
- scikit-learn ships no categorical, Hamming, overlap, Gower or ICM kernel — the full
  `sklearn.gaussian_process.kernels` list is `CompoundKernel, ConstantKernel,
  DotProduct, ExpSineSquared, Exponentiation, Matern, PairwiseKernel, Product, RBF,
  RationalQuadratic, Sum, WhiteKernel`. Checked against the API index on this run. This
  is *not* an argument that Engin should build one; it is why `D9`'s delegation is the
  right call and why the handback matters.
- **No published prevalence figure exists** for how often bioprocess DoE carries a
  categorical factor, and this is `D15`'s burden so here is the trail. Searched
  2026-09-11: Gisperg/Klausser, *Bayesian Optimization in Bioprocess Engineering — Where
  Do We Stand Today?*, [doi:10.1002/bit.28960](https://doi.org/10.1002/bit.28960) —
  Table 1 catalogues 23 BO-in-bioprocess applications 2016–2024 and states only the
  generic "Many optimization problems involve mixed types of input variables", with no
  count; Siska et al., *A Guide to Bayesian Optimization in Bioprocess Engineering*,
  [arXiv:2508.10642](https://arxiv.org/abs/2508.10642) / doi:10.1002/bit.70129 — §5.3
  covers discrete variables, no frequency claim; Desimpel et al., *Chem. Soc. Rev.*
  [doi:10.1039/D5CS00962F](https://doi.org/10.1039/D5CS00962F) — asserts mixed spaces
  are common in chemistry, gives no numbers and covers no fermentation. So the
  prevalence sentence above is built from the three screening papers and is a
  three-paper observation, not a field statistic. Do not promote it to one.

## What it would take

Smallest version that settles it is one paragraph of prose, not code: say in
`ecosystem.md`'s BO slot and in `AGENTS.md` that the estimator-agnostic promise covers
conformal and acquisition and **stops at `CostModel`**, because the default cost model
reads the five simulator knobs positionally. That converts a promise that is currently
three-quarters true into an accurate one, and it is the prerequisite for deciding
anything else.

If the answer is that the seam should hold, the shape is small and does not touch the
GP. `design_context` needs the five physical knobs, and nothing else in
`ParametricCostModel` cares what other columns rode along. Accepting an index or name
map — which columns of `U` are the process knobs — and slicing before
`unit_to_physical` would make the cost head take an augmented matrix. That is a
signature change on one function plus a `knob_columns` field, and it stays a callable
(`D14`). `recommend_batch_by_cost`'s pool sampler is the harder half and should be
scoped separately: sampling a categorical block correctly means enumerating levels, and
at ≤20 levels enumeration beats anything cleverer.

What it does **not** need is a categorical kernel in `engin-core`. `D9` is right here.

## Decisions touched

`D9` — implements, and tests it. The delegation is correct; what is missing is the
return path, which is Engin's own code. `D13` — neutral on the objective, observes the
objective is unreachable for one class of campaign. `D14` — a constraint on any fix:
`knob_columns` as an argument, not a campaign object. `D11` — neutral; the endpoint
path already carries a categorical column. Related and not duplicated:
[#363](https://github.com/enginbio/engin-suite/issues/363) is the sibling type argument
one stage up (`HostQuery.hard` cannot express categorical feasibility);
[#357](https://github.com/enginbio/engin-suite/issues/357) mentions `fit_gp`'s
inability in a validation frame; [#290](https://github.com/enginbio/engin-suite/issues/290)
is the other "a dominant term has no levers" finding and is about downstream structure,
not design-space type.

## Why this might be wrong

**The skeptic pass killed most of what I started with, and the residue is much smaller
than the draft.** I began with "nothing in the suite can represent a qualitative design
factor anywhere." That is flatly false and the counterexample is in the repository:
`engin-protein`'s entire design space *is* categorical — amino-acid identity, 20 levels
per position — one-hot encoded (`featurize.py:96-115`), ridge-fitted
(`lown.py:34-51`), conformally calibrated, and batch-recommended with a Hamming
diversity filter (`lown.py:73-92`). A complete categorical design loop already ships.
I also had ingest as a barrier and it is not, for the reason given above. What is left
is one function and one sampler.

**The strongest surviving counterargument is that `ParametricCostModel` is a
caricature and this is decoration.** Its own docstring calls it "a transparent
three-term caricature", `docs/limitations.md` records that raw material lands near 2%
of modelled cost and capital reads zero without a `ProductionScale`, and the real path
is `BioSteamCostModel` — which is shape-agnostic and so does not have this defect at
all. Widening a seam on a model nobody should use for real numbers is arguably work
spent on the wrong object. The counter is that the caricature is the *default* path,
it is what `recommend_batch_by_cost` uses, and the promise in `AGENTS.md` is made
without qualification — so the prose fix is owed regardless of whether the code fix is.

**Second: the workaround may be good enough, and it is what the field does.** Run one
campaign per level, set `substrate_usd_per_kg` per campaign, compare the $/kg. Every
Engin output is available that way. What you lose is identity *inside* one acquisition
and cost loop — worth 33–50% predictive error in Narayanan's own measurement, but that
is a statement about their surrogate, not about cost, and nobody has shown the cost
ranking changes. I have no evidence that it does.

**Third: prevalence is three papers.** The screening pattern is real and I verified all
three, but three Chinese-language-lab RSM papers from 2025 are not a survey, and the
one actual survey of BO in bioprocess engineering does not report the number. If
someone counts Gisperg's Table 1 properly and it comes back at 2 of 23, the "most
campaigns have one" reading is wrong and the motivation weakens to "some campaigns have
one, and for those the cost head is unreachable" — which is still true and still worth
the prose fix.
